common/djangoapps/entitlements/models.py

```diff
@@ -336,45 +347,59 @@ def get_fullfillable_entitlement_for_user(cls, user, course_run_key):
             CourseEntitlement: The most recent fullfillale CourseEntitlement, None otherwise.
         """
         # Check if the User has any Entitlements that are active
-        if cls.does_user_have_active_entitlements(user):
+        entitlements = cls.get_fullfillable_entitlements(user)
+        if entitlements:
             course_uuid = get_course_uuid_for_course(course_run_key)
             if course_uuid:
-                # Look for an entitlement for the current uuid
-                try:
-                    course_entitlement = cls.objects.filter(
-                        user=user,
-                        course_uuid=course_uuid
-                    ).exclude(
-                        expired_at__isnull=False,
-                        enrollment_course_run__isnull=False
-                    ).latest('created')
-                except CourseEntitlement.DoesNotExist:
-                    return None
-
-                # Verify that the enrollment can be enrolled in
-                if not is_course_run_entitlement_fullfillable(
-                        course_run_key=course_run_key,
-                        entitlement=course_entitlement
-                ):
-                    return None
-                return course_entitlement
+                for entitlement in entitlements:
+                    if course_uuid == entitlement.course_uuid:
+                        # Verify that the enrollment can be enrolled in
+                        if is_course_run_entitlement_fullfillable(
+                                course_run_key=course_run_key,
+                                entitlement=entitlement
+                        ):
+                            return entitlement
         return None
 
-    @classmethod
-    def check_for_existing_entitlement_and_enroll(cls, user, course_run_key):
-        entitlement = cls.get_fullfillable_entitlement_for_user(user, course_run_key)
-        if entitlement:
-            cls.enroll_user_in_course(user, entitlement, course_run_key)
-
     @classmethod
     @transaction.atomic
     def enroll_user_in_course(cls, user, entitlement, course_run_key):
-        enrollment = CourseEnrollment.enroll(
-            user=user,
-            course_key=course_run_key,
-            mode=entitlement.mode
-        )
+        """
+        Enrolls the user in the Course Run and updates the entitlement with the new Enrollment.
+
+        Returns:
+            bool: True if no errors are raised during the enroll.
```

Comment:
For the sake of abstraction, I would describe this truth value as "True if successfully fulfills given entitlement by enrolling the user in the given course run"